Summer 2025 Redesign

stuartlamour opened the pull request:
Delete me!

stuartlamour committed:
CTP-4710 - use mustache templates

Initial commit - WIP

stuartlamour committed:
WIP stash

stuartlamour committed:
WIP stash

stuartlamour committed:
Initial marking summary & export

stuartlamour committed:
Improve download/export

stuartlamour committed:
Remove submissions dropdown

stuartlamour committed:
Separate download/upload function

leonstr commented on renderers/object_renderer.php:
```diff
@@ -489,64 +489,65 @@ public function render_plagiarism_links($files) {
     protected function render_mod_coursework_coursework(mod_coursework_coursework $coursework) {
 
-        global $CFG, $USER;
+        global $CFG, $USER, $PAGE;
```
`$CFG` not used in this function.

leonstr commented on renderers/object_renderer.php:
```diff
@@ -489,64 +489,65 @@ public function render_plagiarism_links($files) {
+            $template->editurl = $link;
+            if($template->feedback || $template->editurl) {
+                $out .= $this->render_from_template('mod_coursework/general_feedback', $template);
             }
```
If we only need `$template` when there's a comment or a link we could move its initialisation into the `if`.

```
            if($coursework->feedbackcomment || $link) {
                $template = new stdClass();
                $template->feedback = $coursework->feedbackcomment;
                $template->editurl = $link;
                $out .= $this->render_from_template('mod_coursework/general_feedback', $template);
            }

```

leonstr commented on renderers/object_renderer.php:
```diff
@@ -489,64 +489,65 @@ public function render_plagiarism_links($files) {
+            // $template->status = $thing->get_status_text();
+            $template->default = true;
+            $template->date = "24th Dec 3:00pm";
+            $template->mark = '33%';
```
Dummy values.

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1341,7 +1342,38 @@ protected function coursework_deadlines_table(mod_coursework_coursework $coursew
 
-        return $tablehtml;
+        // WIP - copied to get marking guide url.
+        $markingguideurl = 'foo';
```
Dummy value.

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1341,7 +1342,38 @@ protected function coursework_deadlines_table(mod_coursework_coursework $coursew
+        $template = new stdClass();
+        $template->description = format_module_intro('coursework', $coursework, $coursework->get_coursemodule_id());
+        $template->markingguideurl = $markingguideurl;
+        $template->duedate = date('jS M g:ia', $normaldeadline);
```
Should we instead be passing the epoch time value to the template and using [the `{{#userdate}}` helper](https://moodledev.io/docs/4.5/guides/templates#userdate)?

leonstr commented on templates/intro.mustache:
```diff
@@ -0,0 +1,66 @@
+{{#duedate}}
+    <h4 class="h6">Due</h4>
+    {{duedate}}
+{{/duedate}}
```
Should we be using `course/format/templates/local/content/cm/activity.mustache` for this kind of information?

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
     }
 
+    protected function coursework_marking_summary(mod_coursework_coursework $coursework): string {
+        global $USER, $PAGE;
```
Globals not used in function.

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+
+        $template = new stdClass();
+        $template->canmark = $canmark;
+        $template->assessor = []; // Initialize the assessor array
```
Surely this comment is overkill?

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+            ['mod/coursework:addinitialgrade', 'mod/coursework:addagreedgrade', 'mod/coursework:administergrades'],
+            $coursework->get_context()
+        );
+        // Set to false under specific condition.
```
Well yes, but isn't that obvious?

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+            $assessablesubmittedsubmissions = $this->remove_unfinalised_submissions($assessablesubmittedsubmissions);
+            $assessablesubmittedsubmissions = $this->remove_ungradable_submissions($assessablesubmittedsubmissions);
+            $finalgradedsubmissions = $this->removed_final_graded_submissions($assessablesubmittedsubmissions);
+            $gradedcount = count($finalgradedsubmissions);
```
`$gradedcount` doesn't seem to be used.

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1571,135 +1568,149 @@ protected function coursework_marking_summary(mod_coursework_coursework $coursew
+                if (has_any_capability(['mod/coursework:addinitialgrade', 'mod/coursework:administergrades'], $coursework->get_context())) {
+                    $allocatedsubsforinitial = $allocatedsubs; // Use the original set
+                    $allocatedsubsforinitial = $this->remove_final_gradable_submissions($allocatedsubsforinitial);
+                    $needsmarking += count($this->get_assessor_initial_graded_submissions($allocatedsubsforinitial));
```
Couldn't we just have:

```
                    $allocatedsubsforinitial = $this->remove_final_gradable_submissions($allocatedsubs);
                    $needsmarking += count($this->get_assessor_initial_graded_submissions($allocatedsubsforinitial));
```

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1571,135 +1568,149 @@ protected function coursework_marking_summary(mod_coursework_coursework $coursew
+                    ];
+
+                    $stages = $coursework->marking_stages();
+                    foreach ($stages as $stage => $s) {
```
We could say `foreach ($stages as $stage => $unused) {` to make it clearer that we're not using the values.

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1571,135 +1568,149 @@ protected function coursework_marking_summary(mod_coursework_coursework $coursew
+    protected function get_export_upload_links(mod_coursework_coursework $coursework): array {
+        global $PAGE;
+
+        $uid = $PAGE->cm->id;
```
Wouldn't this normally be called `$cmid`?

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1571,135 +1568,149 @@ protected function coursework_marking_summary(mod_coursework_coursework $coursew
-                            'count' => $count,
-                        ];
-                    }
+        // Check user capability, and build actions.
```
I can't see that we're checking user capability action in the ensuing code.

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1571,135 +1568,149 @@ protected function coursework_marking_summary(mod_coursework_coursework $coursew
+                    [
+                        'url' => new moodle_url($pageurl, ['export' => 1]),
+                        'lang' => 'exportfinalgrades',
+                        'cap' => ($can('mod/coursework:viewallgradesatalltimes') && $can('mod/coursework:canexportfinalgrades') && $hasfinalised),
```
I think just:

```
                        'cap' => (has_all_capabilities(['mod/coursework:viewallgradesatalltimes', 'mod/coursework:canexportfinalgrades'], $PAGE->context) && $hasfinalised),
```

is shorter and more readable than:

```
        $can = fn(string $cap) => has_capability($cap, $PAGE->context);
        ⋮
                        'cap' => ($can('mod/coursework:viewallgradesatalltimes') && $can('mod/coursework:canexportfinalgrades') && $hasfinalised),
```

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+
+            if ($finalisedsubmissions && !empty($allsubmissions)) {
+                $url = $pageurl . '&download=1';
+                $links['download_submitted_files_url'] = $url;
```
I don't like single-use variables, I don't think it helps the readability here, we could just have:

```
            $links['download_submitted_files_url'] = $pageurl . '&download=1';
```

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+                has_capability('mod/coursework:canexportfinalgrades', $PAGE->context) &&
+                $coursework->get_finalised_submissions()) {
+                $url = $pageurl . '&export=1';
+                $links['export_final_grades_url'] = $url;
```
Again, we could just have:

```
            $links['export_final_grades_url'] = $pageurl . '&export=1';
```

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+                 has_capability('mod/coursework:administergrades', $PAGE->context)) &&
+                $coursework->get_finalised_submissions()) {
+                $url = $pageurl . '&export_grading_sheet=1';
+                $links['export_grading_sheet_url'] = $url;
```
Again, we could just have:

```
            $links['export_grading_sheet_url'] = $pageurl . '&export_grading_sheet=1';
```

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+                $links['upload_grading_sheet_text'] = get_string('uploadgradingworksheet', 'mod_coursework');
+
+                $url = '/mod/coursework/actions/upload_feedback.php?cmid=' . $PAGE->cm->id;
+                $links['upload_feedback_url'] = $url;
```
Again, we could just have:

```
            $links['upload_feedback_url'] = '/mod/coursework/actions/upload_feedback.php?cmid=' . $PAGE->cm->id;
```

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+                 has_capability('mod/coursework:addagreedgrade', $PAGE->context) ||
+                 has_capability('mod/coursework:addallocatedagreedgrade', $PAGE->context) ||
+                 has_capability('mod/coursework:administergrades', $PAGE->context)) &&
+                $coursework->get_finalised_submissions()) {
```
We could have:

```
        if (!empty($allsubmissions) &&
                has_any_capability(['mod/coursework:addinitialgrade',
                     'mod/coursework:addagreedgrade',
                     'mod/coursework:addallocatedagreedgrade'
                     'mod/coursework:administergrades',
                ], $PAGE->context)) &&
                $coursework->get_finalised_submissions()) {
```

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1634,12 +1765,53 @@ protected function coursework_grading_summary_table(mod_coursework_coursework $c
+            has_capability('mod/coursework:addagreedgrade', $PAGE->context) ||
+            has_capability('mod/coursework:addallocatedagreedgrade', $PAGE->context) ||
+            has_capability('mod/coursework:administergrades', $PAGE->context)) &&
+            $coursework->get_finalised_submissions()) {
```
Again, we could use `has_any_capability()`.

leonstr commented on renderers/object_renderer.php:
```diff
@@ -1532,13 +1532,144 @@ private function render_personal_deadline_table_row($personaldeadlinerow) {
+
+            // For users who can add agreed grades or administer grades (or a combination)
+            if (has_any_capability(['mod/coursework:addagreedgrade', 'mod/coursework:administergrades'], $coursework->get_context())
+                || (has_capability('mod/coursework:addinitialgrade', $coursework->get_context()) && has_capability('mod/coursework:addallocatedagreedgrade', $coursework->get_context()))) {
```
It would be shorter to use `has_all_capabilities()`:

```
            if (has_any_capability(['mod/coursework:addagreedgrade', 'mod/coursework:administergrades'], $coursework->get_context())
                        || has_all_capabilities(['mod/coursework:addinitialgrade', 'mod/coursework:addallocatedagreedgrade'], $coursework->get_context())) {
```

leonstr reviewed (changes requested):
This is still in active development so review comments may not be relevant, please ignore as appropriate.

Presumably at some point the UI/UX changes will get reflected in the automated tests ahead of merging.

stuartlamour committed:
WIP - student overview

stuartlamour committed:
WIP - add/edit submission btn

stuartlamour committed:
WIP - submission buttons

stuartlamour committed:
WIP - finalise submission area

stuartlamour committed:
move css to styles.css

stuartlamour committed:
use success class for finalise button

stuartlamour committed:
submission tidy up

stuartlamour committed:
WIP - test dump plagarism

stuartlamour committed:
Update templates

stuartlamour committed:
Move overview col

stuartlamour committed:
border for marking agreed mark

leonstr committed:
Update deadline_extension test for new interface

leonstr committed:
Update factory.feature test for new interface

leonstr committed:
Fixes for feedback_general_feedback Behat test

leonstr committed:
Update grading_summary_visibility for new UI

leonstr committed:
Update submissions_agree_submission_terms for UI

leonstr committed:
Fixes for submissions_deadlines Behat test

stuartlamour committed:
intro mustache update

stuartlamour committed:
CTP-4833 - Marking modal

Initial mustache, css and php.

leonstr committed:
Fixes for submissions_early_finalisation test

Also, added some complementary test scenarios ("should see" to match
"should not see") since one can't be sure if the "should not see" is
working if the HTML for the element has changed.

Also, fixed broken/commented out test scenario.

Also, updated mod_coursework_behat_coursework_page->confirm() to specify
type="submit" as matching all input elements on a page containing <input
type="radio"> appears to cause Behat\Mink\Element\Element->getText() to
fail with 'no such element: Unable to locate element:
{"method":"xpath","selector":"ancestor::form"}'.

stuartlamour committed:
update submission status to allow html

stuartlamour committed:
Merge branch 'CTP-4710---use-mustache-templates' of https://github.com/ucl-isd/moodle-mod_coursework into CTP-4710---use-mustache-templates

stuartlamour committed:
CTP-4833 - Set $submission var in edit_feedback_page

leonstr committed:
Fix: The deadline has now passed...

The message "The deadline has now passed and late submissions (or
updates to existing submissions) are not allowed" always appeared if
there was a due date, even if that date had not been reached.

leonstr committed:
Revert changes to submissions_deadlines

Revert f93dd65's changes to submissions_deadlines.feature and update
mod_coursework_behat_coursework_page->individual_feedback_date_present()
and general_feedback_date_present() instead so this test's custom steps
work.

stuartlamour committed:
CTP-4746 - Dates mustache

stuartlamour committed:
Merge branch 'CTP-4710---use-mustache-templates' of https://github.com/ucl-isd/moodle-mod_coursework into CTP-4710---use-mustache-templates

stuartlamour committed:
CTP-4746 - update intro mustache

stuartlamour committed:
CTP-4746 - update student overview

stuartlamour committed:
CTP-4833 lastmarked

stuartlamour committed:
Use badges for status

stuartlamour committed:
Move general feedback date to general feedback mustache

stuartlamour committed:
Move feedback from submissions to intro mustache

Moved existing_feedback_from_teachers to support mustache output

stuartlamour committed:
Improve object renderer and templates

stuartlamour committed:
Update intro.mustache

stuartlamour committed:
More rubric tables css

stuartlamour committed:
Use lang strings in intro mustache

opitz committed:
Merge pull request #56 from ucl-isd/CTP-4162-moderated-marking

CTP-4162 moderated marking:  refactored $stage_identifier with $stageidentifier.
Addressed dynamic property issues.
Added Behat test.

opitz committed:
Merge pull request #57 from ucl-isd/CTP-4163-view-moderation

CTP-4163: using new API functions to get fullname

leonstr committed:
Add tests for dates at top of page

stuartlamour committed:
Initial feedback display - object_render and intro mustache

1. Add WIP feedback view.
2. Separate intro dates in coursework_into

leonstr committed:
Show elapsed time for overdue submissions

If the submission is late show an elapsed time like "5 hours, 37 minutes
late".

stuartlamour committed:
Feedback add initial css

stuartlamour committed:
Add WIP feedback mustache

leonstr committed:
Improve elapsed time for overdue submissions code

leonstr committed:
Spelling fix

stuartlamour committed:
CTP-4713 - Update feedback display

stuartlamour committed:
CTP-4713 - update feedback

Change render_feedback to public

stuartlamour committed:
CTP-4713 - check markerid is set

stuartlamour committed:
CTP-4713 - check advancedgradinghtml before setting hasadvancedgrading

stuartlamour committed:
CTP-4710 - sort out display for hidden marker

stuartlamour committed:
CTP-4713 - check markerid is set

leonstr committed:
Update Behat test to check student overview output

Also, update Mustache template to use language strings.

stuartlamour committed:
WIP - add initial styles for rubric

stuartlamour committed:
Merge branch 'CTP-4710---use-mustache-templates' of https://github.com/ucl-isd/moodle-mod_coursework into CTP-4710---use-mustache-templates

stuartlamour committed:
Rubric css

stuartlamour committed:
rubric font size css

stuartlamour committed:
Rubric - remove borders, increase font size.

leonstr committed:
Update dates at top of page

Tidy up of deadline, personal deadline and extension dates. Re-add
default personal deadline message for teachers. Add/update Behat tests
for these dates.

stuartlamour committed:
CTP-4713 - Update feedback output

stuartlamour committed:
CTP-4713 - yet more css to overwrite rubric css

stuartlamour committed:
CTP-4713 - Show radio buttons in rubric

stuartlamour committed:
CTP-4757 -Submissions table redesign

aydevworks committed:
CTP-4902 Student column in submission table

aydevworks committed:
CTP-4903 Submission column in submission table

leonstr committed:
Update rendering of student submission status

Work in progress for CTP-4859, includes tests for various submissions
status messages student can see.

stuartlamour commented on lang/en/coursework.php:
```diff
@@ -582,7 +582,7 @@
 $string['notinsitsfeed'] = 'You are not sitting the module that this coursework instance is assessing. Please have another look for the one that corresponds to your module code.';
 $string['notmoderated'] = 'Not moderated';
-$string['notpublishedyet'] = 'Your grade has not been released yet.';
+$string['notpublishedyet'] = 'Your grade has not been released yet';
```
@leonstr Could be a separate ticket, but i've trying to update all the lang strings to use mark instead of grade as we go along.

leonstr replied:
I've changed this, I'll try to watch out for others as I go.

stuartlamour commented on templates/submission.mustache:
```diff
@@ -46,25 +46,48 @@
         {{#date}}
         <li class="d-flex py-1 border-top">
-            {{#str}} tableheadtime, mod_coursework {{/str}} <span class="ml-auto text-right">{{#late}}<span class="badge badge-warning">Late</span>{{/late}}{{#userdate}}{{date}},{{#str}}strftimedatetime, langconfig{{/str}}{{/userdate}}{{#late}}<br> {{.}}{{/late}}</span>
+            {{#str}} tableheadtime, mod_coursework {{/str}} <span class="ml-auto text-right">{{#late}}<span class="badge badge-warning">{{#str}} late, mod_coursework {{/str}}</span>{{/late}}{{#userdate}}{{date}},{{#str}}strftimedatetime, langconfig{{/str}}{{/userdate}}{{#late}}<br> {{.}}{{/late}}</span>
```
@leonstr can this be formatted nicer so its more readable or is that an issue for behat?

leonstr replied:
I did deliberately scrunch this up for Behat (`I should see` doesn't normalise whitespace) but that's not really a good enough reason to have such unreadable mark-up. So I've updated the Behat and wrapped this into multiple lines to hopefully improve this.

stuartlamour commented on templates/submission.mustache:
```diff
@@ -46,25 +46,48 @@
+                        {{status}}
+                    {{/badge}}
+                    {{#stilleditable}}
+                        <br />{{#str}} stilleditable, mod_coursework {{/str}}
```
@leonstr for self closing tags 
```
<hr> <img> 
```
etc the /> isn't necessary anymore in html 5, and it is converted to 
```
<br>
```
 in the browser.

leonstr replied:
Good point, you can tell who wasted their time learning XHTML 20 years ago. I've updated this and I'll try to watch for this in future.

leonstr committed:
CTP-4859 Updates to reflect code review feedback

leonstr committed:
CTP-4859 Update render student submission status

Students now see only four submission statuses: "Not submitted",
"Submitted", "In marking" or "Released". Teachers still see more
detailed statuses, for example, "Graded by some of the initial
assessors").

leonstr committed:
CTP-4652 Remove plagiarism notices on view page

Plagiarism plugin messages -- any disclaimer/agreement, and the EULA if
not agreed -- now appear on the upload page only.

leonstr committed:
CTP-4856 Trivial Behat test for custom description

Also partial revert of f67154c6 to use fixed "I should see the title on
the coursework on the page".

stuartlamour committed:
Try removing current css

stuartlamour committed:
CTP-4979 - use hide instead of disable in settings form

aydevworks committed:
CTP-4904 Marking column in submission table

leonstr committed:
CTP-4857 Update Download and Upload buttons

Buttons now only show when there are submissions.

Changed the string for "Download" -> "Export final grades" to "Export
grades" as there didn’t seem to be anything specific about grade state
affecting what gets exported.

Simplified download URLs changing these from, for example,
/mod/coursework/view.php?id=70&amp;sortby=lastname&amp;sorthow=ASC&amp;
per_page=10&amp;group=-1&download=1 to
/mod/coursework/view.php?id=70&download=1 as the other query string
fields are extraneous.

leonstr committed:
CTP-4159 Change export language strings

Also, remove code for download/export links obviated by new UI.

stuartlamour commented on lang/en/coursework.php:
```diff
@@ -279,8 +279,8 @@
 $string['eventassessableuploaded'] = 'A file has been uploaded.';
-$string['exportfinalgrades'] = 'Export grades';
-$string['exportgradingsheets'] = 'Export grading sheet';
+$string['exportfinalgrades'] = 'Download grades';
```
@leonstr  i'd probably change the string name too, just to help our future selves be less confused.

leonstr replied:
> @leonstr i'd probably change the string name too, just to help our future selves be less confused.

Makes sense, I'll follow this up. I can see other uses of these strings, I need to check if they're redundant, or can be changed, or need to stay the same.

leonstr committed:
CTP-4159 Fix Undefined property $csv_cells

Previously using "Export grading sheet" resulted in:

  Warning: Undefined property: mod_coursework\export\grading_sheet::
  $csv_cells in /mod/coursework/classes/export/grading_sheet.php on line
  137
  Warning: foreach() argument must be of type array|object, null given in
  /mod/coursework/classes/export/csv.php on line 115

leonstr committed:
Fix unexpected Behat failures

Behat tests have been unexpectedly failing with Chrome 138.0.7204.xxx.
Switching to Firefox.

leonstr committed:
CTP-4857 Change lang string and string IDs

Update language strings as requested during testing and code review.

coursework.js was using <input type="hidden" id="element_lang_messages">
to provide access to one of these strings (among others) but it did not
appear to be actually used. This element has been removed and the only
string being used ('loadingpagination') is now loaded with core/str.

leonstr committed:
CTP-4857 Update Behat test for string change

aydevworks committed:
CTP-4905 Action column

aydevworks committed:
Use modal for finalise and unfinalise submission

aydevworks committed:
CTP-5017 Display group submission in submission table

leonstr committed:
CTP-4855 Marking summary

Code tidy-up for the "Marking summary" table for the revised user
interface.

The isfinalgrade = 1 was removed from there_is_final_feedback() as this
causes incorrect marking summary totals (as far as I can see
isfinalgrade is always 0 when mod_coursework is used through the web
interface).

Test marking_summary_double.feature fails due to an issue introduced in
an earlier commit (see CTP-5019).

aydevworks committed:
CTP-5019 Fix undefined property warning

aydevworks committed:
CTP-4961 Implement release marks button

leonstr committed:
CTP-4858 General feedback

Code tidy-up for the "General feedback" field for the revised user
interface.

Also, refactor Mustache templates and renderer function for intro area
so that <hr> dividers are only shown when there's content to separate.

Also, fix existing issue where if "General feedback release date" is
set the general feedback is never shown to the student.

watson8 committed:
CTP-5002 avoid mod_assign constant (#61)

* CTP-5002 avoid mod_assign constant

* CTP-5002 behat test zip

* CTP-5002 add upload behat zip

watson8 committed:
CTP-4916 due dates in course (#64)

leonstr committed:
CTP-4795 Student feedback view

Code tidy-up for the student's view of the feedback from markers.

Also, show feedback date for anonymous feedback.

Also, removed namespaced-out CSS (.path-mod-foo)

Also, removed some functions for Behat custom steps where the
corresponding interface items are no longer present.

aydevworks committed:
CTP-4529 Rename submitted files using student candidate number

watson8 committed:
CTP-5043 marking guide agreed form (#67)
